apps/gitness/src/pages/pull-request/context/pull-request-data-provider.tsx

```diff
@@ -299,161 +121,124 @@ const PullRequestDataProvider: React.FC<PullRequestDataProviderProps> = ({ child
-      })
-        .then(({ body: res }) => {
-          // if (isMounted.current) {
-
-          if (res?.rule_violations?.length && res?.rule_violations?.length > 0) {
-            setRuleViolation(true)
-            setRuleViolationArr({ data: { rule_violations: res?.rule_violations } })
-            // TODO: fix allowed strats to work with this structure
-            // setAllowedStrats(res.allowed_methods)
-            setRequiresCommentApproval?.(res.requires_comment_resolution ?? false)
-            setAtLeastOneReviewerRule?.(res.requires_no_change_requests ?? false)
-            setReqCodeOwnerApproval?.(res.requires_code_owners_approval ?? false)
-            setMinApproval?.(res.minimum_required_approvals_count ?? 0)
-            setReqCodeOwnerLatestApproval?.(res.requires_code_owners_approval_latest ?? false)
-            setMinReqLatestApproval?.(res.minimum_required_approvals_count_latest ?? 0)
-            setConflictingFiles?.(res.conflict_files)
-          } else {
-            setRuleViolation(false)
-            // TODO: fix allowed strats to work with this structure
-            // setAllowedStrats(res.allowed_methods)
-            setRequiresCommentApproval?.(res.requires_comment_resolution ?? false)
-            setAtLeastOneReviewerRule?.(res.requires_no_change_requests ?? false)
-            setReqCodeOwnerApproval?.(res.requires_code_owners_approval ?? false)
-            setMinApproval?.(res.minimum_required_approvals_count ?? 0)
-            setReqCodeOwnerLatestApproval?.(res.requires_code_owners_approval_latest ?? false)
-            setMinReqLatestApproval?.(res.minimum_required_approvals_count_latest ?? 0)
-            setConflictingFiles?.(res.conflict_files)
-          }
-        })
-        .catch(err => {
-          // if (isMounted.current) {
-          if (err.status === 422) {
-            setRuleViolation(true)
-            setRuleViolationArr(err)
-            // setAllowedStrats(err.allowed_methods)
-            setRequiresCommentApproval?.(err.requires_comment_resolution)
-            setAtLeastOneReviewerRule?.(err.requires_no_change_requests)
-            setReqCodeOwnerApproval?.(err.requires_code_owners_approval)
-            setMinApproval?.(err.minimum_required_approvals_count)
-            setReqCodeOwnerLatestApproval?.(err.requires_code_owners_approval_latest)
-            setMinReqLatestApproval?.(err.minimum_required_approvals_count_latest)
-            setConflictingFiles?.(err.conflict_files)
-          } else if (
-            err.status === 400
-            // &&
-            // (getErrorMessage(err) === oldCommitRefetchRequired || getErrorMessage(err) === prMergedRefetchRequired)
-          ) {
+  useEffect(() => {
+    const hasChanges =
+      !isEqual(store.pullReqMetadata, pullReqData) ||
+      !isEqual(store.pullReqStats, pullReqData?.stats) ||
+      !isEqual(store.pullReqCommits, commits) ||
+      !isEqual(store.pullReqActivities, activities)
```

Comment:
I was going to say that since we're using `immer`, it should be impossible for the objects to change without the references to those objects being different, so you might not need `isEqual`, but since we're using `immer` it'll start off by duplicating the object (and the amount of memory it was occupying) and thus breaking the ability to just check for simple exactness.